File: trees/199.py
# Level Order Traversal(BFS)
# Tc = O(n)
# Sc = O(n)
from collections import deque
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.right = right
        self.left = left

class Solution:
    def rightSideView(self, root: TreeNode) -> list[int]:
        if root is None:
            return []
        result = []
        que = deque([root])
        while que:
            n = len(que)
            node = None
            for _ in range(n):
                node = que.popleft()
                if node.left:
                    que.append(node.left)
                if node.right:
                    que.append(node.right)
            result.append(node.val)
        return result

# Pre Order Traversal(DFS)
# Tree level(eg.3) = output level(eg.3) bkz each level first element will be seen from right side view
class Solution:
    def preOrder(self, root, level, result):
        if root is None:
            return
        if len(result) < level:
            result.append(root.val)
        self.preOrder(root.right, level + 1, result)
        self.preOrder(root.left, level + 1, result)

    def rightSideView(self, root):
        result = []
        self.preOrder(root, 1, result)
        return result

File: trees/199.js
class TreeNode {
    constructor(val=0, left=null, right= null){
        this.val = val
        this.left = left
        this.right = right
    }
}
function rightSideView(root){
    if(!root) return []
    result = []
    const queue = [root]
    while(queue.length >0){
        const levelSize = queue.length
        let rightNode = null
        for(let i = 0; i<levelSize; i++){
            rightNode = queue.shift()
            if(rightNode.left) queue.push(rightNode.left)
            if(rightNode.right) queue.push(rightNode.right)
        }
    if(rightNode){
        result.push(rightNode.val)
    }
    }
    return result
}


// Pre-Order traversal(DFS)
class Solution{
    preOrder(root, level, result){
        if(root === null)
            return;
        if(result.length < level){
            result.push(root.val)
        }
        this.preOrder(root.right, level + 1,result)
        this.preOrder(root.left, level + 1, result)
    }
    rightSideView(root){
        const result = []
        this.preOrder(root, 1, result)
        return result
    }
}